hail/python/test/hailtop/inter_cloud/test_fs.py

```diff
@@ -630,3 +631,51 @@ async def test_rmtree_on_symlink_to_directory():
         finally:
             await fs.rmtree(sema, str(base))
             assert not await fs.isdir(str(base))
+
+
+async def test_operations_on_a_bucket_url_is_error(filesystem: Tuple[asyncio.Semaphore, AsyncFS, AsyncFSURL]):
+    _, fs, base = filesystem
+
+    if base.scheme in ('', 'file'):
+        return
+
+    bucket_url = str(base.with_path(''))
+
+    with pytest.raises(IsABucketError):
+        await fs.isdir(bucket_url)
+
+    assert await fs.isfile(bucket_url) is False
+
+    with pytest.raises(IsABucketError):
+        await fs.statfile(bucket_url)
+
+    with pytest.raises(IsABucketError):
+        await fs.remove(bucket_url)
+
+    with pytest.raises(IsABucketError):
+        await fs.create(bucket_url)
+
+    with pytest.raises(IsABucketError):
+        await fs.open(bucket_url)
+
+
+async def test_hfs_ls_bucket_url_not_an_error(filesystem: Tuple[asyncio.Semaphore, AsyncFS, AsyncFSURL]):
+    _, _, base = filesystem
+
+    if base.scheme in ('', 'file'):
+        return
+
+    bucket_url = str(base.with_path(''))
+    fs = RouterFS()
```

Comment:
this resource needs to be closed

Reply:
I made RouterFS a context manager and used that here. I also added an exit stack to LocalBackend. I also changed ServiceBackend to use its exitstack the same way as LocalBackend.